src/rt/sched/work.h

```diff
@@ -0,0 +1,126 @@
+// Copyright Microsoft and Project Verona Contributors.
+// SPDX-License-Identifier: MIT
+#pragma once
+
+#include "epoch.h"
+
+#include <atomic>
+#include <snmalloc/snmalloc.h>
+
+namespace verona::rt
+{
+  using namespace snmalloc;
+
+  /**
+   * @brief A work item that can be scheduled.
+   *
+   * The construction of this is handled by other types that embed
+   * this, and the code pointer.  See Closure below for an example.
+   */
+  struct Work
+  {
+    static constexpr auto NO_EPOCH_SET = (std::numeric_limits<uint64_t>::max)();
+
+    // Entry in the MPMC Queue of work items per scheduler.
+    union
+    {
+      // When in the queue this references the next element
+      std::atomic<Work*> next_in_queue;
+      // When not in the queue this is the epoch that it was last seen in the
+      // queue. This is used to allow deferred deallocation to be more
+      // efficient.
+      uint64_t epoch_when_popped{NO_EPOCH_SET};
+    };
+
+    // The function to execute this work item. It is supplied with a self
+    // pointer and is responsible for all casting and memory management.
+    void (*f)(Work*);
+
+    constexpr Work(void (*f)(Work*)) : f(f) {}
+
+    // Helper to run the item.
+    void run()
+    {
+      f(this);
+    }
+
+    /**
+     * Helper to perform deallocation.
+     *
+     * Due to complexity with the MPMCQ memory management there is an
+     * epoch mechanism to prevent the underlying allocator decommitting
+     * the memory and thus causing a segfault.  This correctly deallocates
+     * a work item.
+     *
+     * It assumes the work item is the start of an underlying allocation, and
+     * does not work if it is embedded not at the start of an allocation.
+     */
+    void dealloc()
+    {
+      auto& alloc = ThreadAlloc::get();
+      auto epoch = epoch_when_popped;
+      auto outdated = epoch == NO_EPOCH_SET || GlobalEpoch::is_outdated(epoch);
+      if (outdated)
+      {
+        Logging::cout() << "Work " << this << " dealloc" << Logging::endl;
+        alloc.dealloc(this);
+      }
+      else
+      {
+        Logging::cout() << "Work " << this << " defer dealloc" << Logging::endl;
+        // There could be an ABA problem if we reuse this work as the epoch
+        // has not progressed enough We delay the deallocation until the epoch
+        // has progressed enough
+        // TODO: We are waiting too long as this is inserting in the current
+        // epoch, and not `epoch` which is all that is required.
+        Epoch e(alloc);
+        e.delete_in_epoch(this);
+      }
+    }
+  };
+
+  /**
+   * Builds a work item from a C++ type, T.
+   *
+   * The internal layout is done manually to ensure we
+   * aren't doing anything UB.
+   */
+  class Closure
```

Comment:
Sorry about being dense, but would a helper class that is a sub-type of Work and with a single member of type T help simplify some of this logic and avoid needing pointer offsets?

Reply:
Yes if the subclass is a [standard layout class](https://en.cppreference.com/w/cpp/language/classes#Standard-layout_class).  However, my interpretation of standard layout means almost nothing is one (can't use public/private or static).  If it isn't a standard layout, then a bunch of things like `offsetof` and `CONTAIN_RECORD` are all UB. So the casting required for dispatching from Work to Closure becomes UB, and I am very concerned about it.

That said, I think it would be okay in all compilers, but probably not due to the language standard.  I may also be interpreting things overly cautiously, but this is a pretty core part of the runtime, so I want to make sure it is correct.

Effectively, I have had to implement my own layout.

Reply:
Thanks, this makes sense!